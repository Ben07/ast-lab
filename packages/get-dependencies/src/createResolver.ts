import enhancedResolve, { ResolverFactory } from 'enhanced-resolve';
import path from 'path';

import isCore from './isCore';
import isThirdParty from './isThirdParty';

export default function createResolver(
  opts: ResolverFactory.ResolverOption = {}
) {
  return async (mod: string, source: string = '/'): Promise<string | void> => {
    if (!mod || isCore(mod)) {
      return;
    }
    if (path.isAbsolute(mod)) {
      return mod;
    }
    const basedir = path.dirname(source || '/');
    const result = await new Promise((resolve, reject) => 
      enhancedResolve.create(opts)(
        basedir, 
        mod, 
        (err: Error, res: string | undefined) => {
          if (err) {
            return reject(err.message);
          }
          if (!res) {
            console.log(`Couldn't find module ${mod}!`);
            return resolve();
          }
          if (isThirdParty(res)) {
            return resolve();
          }
          return resolve(res);
        }
      )
    ) as string | void;
    return result;
  }
}
