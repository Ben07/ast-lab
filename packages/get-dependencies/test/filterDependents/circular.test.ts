// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import { filterDependents } from './utils';
import createResolver from '../../src/createResolver';

enumerateArgsTestFunction(
  filterDependents.bind(null, 'circular'),
  configArgs()
  .arg('sources', [
    ['a.js', 'b.js', 'c.js'],
  ])
  .arg('targets', [
    { './a': ['default'] }
  ])
  .arg('options', [
    {
      resolver: createResolver({
        modules: ['../']
      })
    }
  ]),
  'circular'
);