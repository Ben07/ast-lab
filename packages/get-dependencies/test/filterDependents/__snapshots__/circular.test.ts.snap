// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circular - (["sources", "targets", "options"]) - [["a.js","b.js","c.js"],{"./a":["default"]},{}] 1`] = `
Array [
  "/circular/a.js",
  "/circular/b.js",
  "/circular/c.js",
]
`;
